File: Data Structures & Algorithms/meeting-schedule-ii/submission-1.py
"""
Definition of Interval:
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""
from collections import deque
from heapq import heappop, heappush

class Solution:
    def minMeetingRooms(self, intervals: List[Interval]) -> int:
        if len(intervals) <= 1: return len(intervals)
        intervals = sorted(intervals, key=lambda interval: interval.start)

        curr = 1
        res = 1
        rooms = [intervals[0].end]
        print(intervals)
        for interval in intervals[1:]:
            print(interval, rooms, curr, res)
            if interval.start > rooms[0]:
                heappop(rooms)
                curr -= 1
            else:
                res = max(curr, res)
            curr += 1
            heappush(rooms, interval.end)
            print(interval, rooms, curr, res)

        return max(curr, res)

File: Data Structures & Algorithms/trapping-rain-water/submission-1.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if (height.size() < 2) return 0;
        int l = 0;
        int r = height.size() - 1;
        int maxl = height[l];
        int maxr = height[r];
        int res = 0;
        while (l < r) {
            if (maxl < maxr) {
                l++;
                maxl = max(height[l], maxl);
                res += (maxl - height[l]);
            } else {
                r--;
                maxr = max(height[r], maxr);
                res += (maxr  -height[r]);
            }
        }
        return res;
    }
};
